### PR TITLE
Teacher Feedback - Dependent destroy

### PR DESCRIPTION
**Pull request by epeach**
Per discussions with Poorva, if a teacher is deleted, we want to delete the feedback they have provided.

Note: Tests incoming

### CODE REVIEW

**Comment by epeach** [comment]
@aoby - Do I need to specify in user.rb that, the dependent destroy should only happen if the user is a teacher (since technically both student-users and teacher-users have teacher_feedbacks)?

**Comment by aoby** [comment]
Yes, in a way. The association itself needs to be only based on the [teacher_id](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/teacher_feedback.rb#L9) in TeacherFeedbacks, not the `student_id`. There is nothing special to add to the `dependent_destroy` but the association needs to be fixed. By default the `has_many` association assumes the foreign key in the associated model is based on the class name where it's defined. In other words, this is looking for `teacher_feedbacks.user_id`, which does not exist.

It should be defined as:
```ruby
has_many :teacher_feedbacks, foreign_key: 'teacher_id', dependent: :destroy
```

And add a unit test to make sure it works.

**Comment by aoby** [comment]
Also if you wanted both, you could add two `has_many` associations, one for teacher and one for student.

Something like:
```ruby
has_many :teacher_feedbacks, foreign_key: 'teacher_id', dependent: :destroy
has_many :teacher_feedbacks_as_student, class: 'TeacherFeedback', foreign_key: 'student_id', dependent: :destroy
```

Note you need to specify `class` if it can't be assumed from the association name.

**Comment by epeach** [comment]
@aoby Thanks for the help! As far as I know, we just want dependent delete for teachers.

**Comment by epeach** [comment]
cc @islemaster

File: dashboard/test/models/teacher_feedback_test.rb [hunk 1]
```diff
@@ -62,4 +62,23 @@ class TeacherFeedbackTest < ActiveSupport::TestCase
...
+  end
+
+  test 'does not destroy when student is destroyed' do
+    teacher = create :teacher
+    student = create :student
+    first_feedback = create :teacher_feedback, teacher: teacher, student: student
+    second_feedback = create :teacher_feedback, teacher: teacher, student: student
+    student.destroy
+    assert TeacherFeedback.exists?(first_feedback.id)
+    assert TeacherFeedback.exists?(second_feedback.id)
```
Comment (islemaster) [inline dashboard/test/models/teacher_feedback_test.rb:89]: If you wanted to test that the soft-deleted feedback can be recovered with the user, I believe you'd call
```rb
user.restore(:recursive => true, :recovery_window => 2.minutes)
```
and then verify that the feedback had been restored.

Comment (aoby) [reply dashboard/test/models/teacher_feedback_test.rb:89]: Or call User.undestroy: https://github.com/code-dot-org/code-dot-org/blob/66b8da828d0bc929ff7b13a29d0ea75ac3d402fb/dashboard/app/models/user.rb#L1958-L1963

File: dashboard/test/models/teacher_feedback_test.rb [hunk 2]
```diff
@@ -62,4 +62,28 @@ class TeacherFeedbackTest < ActiveSupport::TestCase
...
+    first_feedback = create :teacher_feedback, teacher: teacher
+    second_feedback = create :teacher_feedback, teacher: teacher
+    teacher.destroy
+    assert_not TeacherFeedback.exists?(first_feedback.id)
+    assert_not TeacherFeedback.exists?(second_feedback.id)
+
+    #Assert that feedback was soft-deleted
+    teacher.undestroy
+    assert TeacherFeedback.exists?(first_feedback.id)
+    assert TeacherFeedback.exists?(second_feedback.id)
```
Comment (aoby) [inline dashboard/test/models/teacher_feedback_test.rb]: This asserts that the record can be un-deleted. If you just want to verify that it's soft-deleted, you can do something like this:

```ruby
assert first_feedback.reload.deleted?
```
or 
```ruby
assert TeacherFeedback.with_deleted.exists?(first_feedback.id)
```

Comment (aoby) [reply dashboard/test/models/teacher_feedback_test.rb]: Or just replace the 2 assertions on lines 71 and 72 with something more like my first suggestion here, which will confirm that they are indeed (soft) deleted.

**Pull request merged**